libdnf/rpm/solv/map_iterator.hpp

```diff
@@ -0,0 +1,150 @@
+/*
+Copyright (C) 2020 Red Hat, Inc.
+
+This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+
+#ifndef LIBDNF_RPM_SOLV_MAP_ITERATOR_HPP
+#define LIBDNF_RPM_SOLV_MAP_ITERATOR_HPP
+
+
+#include <bits/stdc++.h>
+#include <solv/bitmap.h>
+#include <solv/pooltypes.h>
+
+#include <iterator>
+
+
+namespace libdnf::rpm::solv {
+
+
+class SolvMap;
+
+
+class SolvMapIterator {
+public:
+    explicit SolvMapIterator(const Map * map);
+    SolvMapIterator(const SolvMapIterator & other);
+
+    using iterator_category = std::forward_iterator_tag;
+    using difference_type = Id;
+    using value_type = Id;
+    using pointer = void;
+    using reference = void;
+
+    Id operator*() const { return id; }
+
+    SolvMapIterator & operator++();
+    SolvMapIterator operator++(int) { return ++(*this); }
+
+    bool operator==(const SolvMapIterator & other) const { return id == other.id; }
+    bool operator!=(const SolvMapIterator & other) const { return id != other.id; }
+
+    void begin() { id = BEGIN; }
+    void end() { id = END; }
+
+protected:
+    const Map * get_map() const noexcept { return map; }
+
+private:
+    constexpr static int BEGIN = -1;
+    constexpr static int END = -2;
+
+    // pointer to a map owned by SolvMap
+    const Map * map;
+
+    // current address in the map
+    unsigned char * map_current;
+
+    // Id representing the first bit in the block
+    //Id map_current_id;
+
+    // the last address in the map
+    const unsigned char * map_end;
+
+    // value of the iterator
+    Id id = BEGIN;
```

Comment:
Please could you rename variable id to something that more represents purpose. Like `iterator_position` or anything else.

Reply:
I'll use `iterator_value` or something similar.

Reply:
Replaced with current_value which seems to be better than iterator_value, because it's part of an iterator class and iterator.current_value is better than iterator.iterator_value.